mqtt.go

```diff
@@ -19,598 +24,613 @@ const (
+		// pubAckCh: We will receive MqttPublishID's, which will correspond
+		//   to the return value of Publish()
+		// pingAckCh: True if our ping received an ACK or false if timeout
+		// The delegate is responsible for creating these channels to be of
+		// sufficient size. This is the only mechanism for communicating with
+		// the user. If necessary, the user can create a queue to ensure
+		// that no data is lost.
+		ReceiveChannels() (subRecvCh chan<- MqttSubData,
+			queueAckCh chan<- MqttQueueResult,
+			pingAckCh chan<- bool)
+
+		// How long to wait for requests to finish
+		RequestTimeout() time.Duration
+
+		// Retrieve the topics for this delegate
+		Subscriptions() []string
+
+		// Hook so we can clean up on closing of connections
+		Close()
+	}
+
+	MqttClient struct {
+		// Provides the public API to MQTT. We handle
+		// connect, disconnect, ping (internal)
+		// publish, and subscribe. All functions are synchronous.
+
+		isConnected bool
+		delegate    MqttDelegate
+		conn        *mqttConn
+		wgSend      sync.WaitGroup // wait on sending
+		wgRecv      sync.WaitGroup // wait on receiving
+	}
+
+	// Delegate for the mqqtConn to call back to the MqttClient
+	mqttPubReceiver interface {
+		handlePubReceive(*packet.PublishPacket)
+	}
+
+	// Type alias for something that we use everywhere
+	mqttPacketChan chan packet.Packet
+	// Internal structure used by the client to communicate with the mqttd
+	// server. This is a thin wrapper over the mqtt_packet package.
+	mqttConn struct {
+		// delegate to handle receiving publishes. Of course, achannel would
+		// be an alternate way to do this, but this delegate is currently
+		// just doing some verification and getting the bytes out of the
+		// packet and putting it into a channel, so adding another channel
+		// for that would add unneeded overhead. If circumstances change,
+		// change, we can revisit this decision.
+		PubHandler mqttPubReceiver
+
+		conn   net.Conn
+		stream *packet.Stream
+		// Sequential packet ID. Used to match acks our actions (pub, sub)
+		// excluding connects and pings. We also don't have packet ID's for
+		// receiving pubs from our subscriptions because we didn't initiate them.
+		lastPacketID uint16
+
+		// Channels to respond to clients
+		packIDToChan map[uint16]mqttPacketChan
+		outPktCh     mqttPacketChan
+		connRespCh   mqttPacketChan
+		pingRespCh   chan<- bool
+		queueAckCh   chan<- MqttQueueResult
+
+		mutex sync.Mutex
+	}
+
+	MqttClientError error
+)
 
-func (mc *mqttConn) sendPacket(p packet.Packet) error {
-	if err := mc.stream.Write(p); err != nil {
-		logError("[MQTT] failed to send %s packet: %s", p.Type(), err.Error())
-		return err
+// Creates a client and connects. A client is invalid if not connected, and
+// you need to create a new client to reconnect and subscribe
+// a pingInterval of zero means no keepalive pings
+func NewMqttClient(mqttHost string, mqttPort int, tlsConfig *tls.Config,
+	useTLS bool, delegate MqttDelegate) *MqttClient {
```

Comment:
Here is why I'm not comfortable MqttDelegate manage username and password. My understanding is that TLS is part of the MQTT authentication mechanism.
http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/csprd02/mqtt-v3.1.1-csprd02.html#_Toc385349852

My suggestion may be easier to understand for who familiar with MQTT.
```suggestion
	useTLS bool, username string, password string, delegate MqttDelegate) *MqttClient {
```

Reply:
I haven't thought much about TLS, so it's a good time to talk about it now. Your API change is good, but the TLS config is in the delegate, since it's not part of the device. What do you think, @takjn

Reply:
Thank you, @shigmas. I understand. How about `AuthInfo()` returns TLS config in addition to username and password?